(Chore) #2254 Show generic message on error

yaroslav-fedyshyn-nordwhale opened the pull request:
# Description

- Made a general error message builder.
- use it in places where e.message shown for user via dialog.
- create excell file with the described places where the message was changed

About #2254 

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

## Excell File:
[GoodDollar Error Codes Accross the Project.xlsx](https://github.com/GoodDollar/GoodDAPP/files/5083442/GoodDollar.Error.Codes.Accross.the.Project.1.xlsx)

yaroslav-fedyshyn-nordwhale commented:
@sirpy I found such exceptions in project:
- EMAIL-UNEXPECTED-1
- CANCEL-PAYMNET-1
- CANCEL-PAYMNET-2

Should I refactor/rename them and show like the other I added in this PR?

sirpy commented on src/components/signup/SignupState.js:
```diff
@@ -486,11 +488,14 @@ const Signup = ({ navigation }: { navigation: any, screenProps: any }) => {
+        dialogShown: true,
+        errorCode: serverResponse && 8,
+      })
+      showSupportDialog(showErrorDialog, hideDialog, navigation.navigate, serverResponse ? message : generalError(8))
```
we need to check what are the possible server messages i'm not sure we want to show them

yaroslav-fedyshyn-nordwhale replied:
@sirpy 
There could be such error messages:
- User mobile not verified!
- User email not verified!
- You cannot create more than 1 account with the same credentials
- Failed adding user to mautic
- Failed adding user to w3
- Failed topping user wallet

Also there is some places where we throw system errors:
```
.catch(e => {
          logger.error('updateMarketToken failed', e.message, e, { userRecordWithPII })
          throw e
        })
```

```
.catch(e => {
            logger.error('addUserToWhiteList failed', e.message, e, { userRecord })
            throw e
          })
```

```
.catch(e => {
          logger.error('failed updating new user record', e.message, e, { toUpdateUser })
          throw e
        })
```

I think we can throw the error with some human-readable error messages here too.
For example:
- Something went wrong while trying to fetch market token
- Something went wrong while trying to white-list user
- Something went wrong while trying update user record

sirpy replied:
lets just stick with the generic error  message

serdiukov-o-nordwhale commented on src/lib/utils/generalError.js:
```diff
@@ -0,0 +1 @@
+export default errorCode => `Sorry, Some unexpected error (E${errorCode}) happened, please try again`
```
1. I moved all error utils onto logger/exceptions.js Plese adjust all imports
2. use decorate() util to add code/name to the exception. Don't report the code with custom log payload. Also this utill will return 'Sorry some unexpected...' string you could show in the dialog to the endusers:

```javascript
import { decorate, ExceptionCode } from 'path/to/logger/exceptions'

const { E1, E2, ... } = ExceptionCode

// ... 
 const { message } = exception
 const uiMessage = decorate(exception, E1)

 // you don't  need to specify the code in the object param - decorate will set exception.code = 1 and exception.name = 'E1'
 log.error('evaluateFRValidity failed', message, exception, { dialogShown: true })
 showErrorDialog(uiMessage(1), '', { // ...
```

3. i see we still have incorrect error logging (e.g. evaluateFRValidity failed - no error object was logged just after the message). Please re check the app very carefully, find & fix all the similar cases

serdiukov-o-nordwhale commented:
Adjusted your PR. Just check pt. 3 from my previous comment and also please check the @sirpy 's comment

yaroslav-fedyshyn-nordwhale commented:
> have we gone through all the places where we show error?

@sirpy 
I double-checked all the places where we show error for the user. I didn't find any missed places.
See my first comment in this PR

yaroslav-fedyshyn-nordwhale commented:
> @yaroslav-fedyshyn-nordwhale
> Adjusted your PR. Just check pt. 3 from my previous comment and also please check the @sirpy 's comment

@serdiukov-o-nordwhale I went through the error logs across the project and verify the argument convention. See my last commit.

serdiukov-o-nordwhale commented on src/lib/API/api.js:
```diff
@@ -425,7 +425,7 @@ export class APIService {
       payload,
       () => log.info('addMauticContact success'),
-      ({ content }, xhr) => log.error('addMauticContact call failed:', content),
+      ({ content }, xhr) => log.error('addMauticContact call failed:', '', null, { content }),
```
Could you check which props also having the object comes from Mautic ? Probably there something like error, message etc so we could create an Error object and log it

yaroslav-fedyshyn-nordwhale replied:
The first argument received in error callback is response object - I can't find some docs/examples of error callback arguments; Also, I tried to emulate errors from mautic and get just empty string. So I will send hole object to the log as extra data

serdiukov-o-nordwhale replied:
ok

serdiukov-o-nordwhale commented on src/lib/gundb/UserStorageClass.js:
```diff
@@ -1237,7 +1237,7 @@ export class UserStorage {
       .get('value')
       .decrypt()
-      .catch(e => logger.error('getProfileFieldValue decrypt failed:', e.msg, e))
+      .catch(e => logger.error('getProfileFieldValue decrypt failed:', e.message || e.msg, e))
```
.decrypt could throw an Error object or just a string. in the case of string let's create an Error object

```suggestion
      .catch(reason => {
        let exception = reason
        let { message } = exception
        
        if (!isError(reason)) {
          message = reason
          exception = new Error(reason)
        }
        
        logger.error('getProfileFieldValue decrypt failed:', message, exception)
      })
```

sirpy commented:
@yaroslav-fedyshyn-nordwhale 
1. yes change them
2. in the signup process and in the profile mobile/email verification we show errors not in dialog but under the input field. we need to take these into account also

yaroslav-fedyshyn-nordwhale commented:
> @yaroslav-fedyshyn-nordwhale
> 
> 1. yes change them
> 2. in the signup process and in the profile mobile/email verification we show errors not in dialog but under the input field. we need to take these into account also

1. I pushed a fix with a human-readable error message for POST /user/add endpoint on the server. Here is the commit hash - e432f497fc541be51b9b230734277cd94bbe80af
2. I add error handling and send human-readable error messages from the server to the client for email/phone verification endpoints. commit hash - 64fc673988d9a943cdd60b140d048072e5d01e5f

yaroslav-fedyshyn-nordwhale commented:
@serdiukov-o-nordwhale please check if I merged the analytic.js patchLogger changes correctly

yaroslav-fedyshyn-nordwhale commented:
I updated the excel file in PR description.

yaroslav-fedyshyn-nordwhale commented:
> 1. refactor EMAIL-UNEXPECTED-1 etc to E codes
> 2. i dont understand where are the commit hashes, they should be in this PR
> 3. stick to the generic error message on the different server errors (in signup finished)

1. Fixed in new commit
2. its on server. I made human-readable error messages to be sent from server in case some unexpected errors occurred.
3. Fixed in new commit.

sirpy commented on src/components/signup/SignupState.js:
```diff
@@ -489,19 +488,13 @@ const Signup = ({ navigation }: { navigation: any, screenProps: any }) => {
       })
 
-      showSupportDialog(showErrorDialog, hideDialog, navigation.navigate, uiMessage)
+      showSupportDialog(showErrorDialog, hideDialog, navigation.navigate)
```
this is still E8

serdiukov-o-nordwhale commented on src/components/dashboard/FeedList.js:
```diff
@@ -128,13 +128,13 @@ const FeedList = ({
               })
               userStorage.updateOTPLEventStatus(id, 'pending')
-              showErrorDialog('The payment could not be canceled at this time', 'CANCEL-PAYMNET-1')
+              showErrorDialog('The payment could not be canceled at this time. Please try again', ExceptionCode.E11)
```
why decorate isn't used ? it adds Exx as the exception.code to be reported to Sentry

please see the other places where decorate is used

yaroslav-fedyshyn-nordwhale replied:
The task requirement was to replace the technical error message with a general human-readable error message which won't confuse the users in places where the e.message is shown. And mark those places with Ex code.
Exactly this place already contain an understandable error message.

yaroslav-fedyshyn-nordwhale replied:
Ok. I see. I used decorate for those places you asked as an error code should be sent to the sentry. See new commit

serdiukov-o-nordwhale commented on src/lib/logger/exceptions.js:
```diff
@@ -7,7 +7,7 @@ export const ExceptionCategory = {
 }
 
-export const ExceptionCode = fromPairs(range(1, 10).map(code => [`E${code}`, code]))
+export const ExceptionCode = fromPairs(range(1, 13).map(code => [`E${code}`, code]))
```
range doesn't includes the upper bound. so for E1-E-13 codes you have to call range(1, 14). 
probably it would be good to suggest the code as the following to make it more clear:

```suggestion
const MAX_EXCEPTION_CODE = 13

export const ExceptionCode = fromPairs(range(1, MAX_EXCEPTION_CODE + 1).map(code => [`E${code}`, code]))
```

serdiukov-o-nordwhale commented on src/components/dashboard/Claim.js:
```diff
@@ -106,9 +107,13 @@ const Claim = props => {
-      log.error('evaluateFRValidity failed', e.message, { dialogShown: true })
-      showErrorDialog('Sorry, Something unexpected happened, please try again', '', {
+    } catch (exception) {
+      const { message } = exception
```
add decorate

yaroslav-fedyshyn-nordwhale replied:
it already there

serdiukov-o-nordwhale commented on src/components/dashboard/FeedList.js:
```diff
@@ -127,13 +128,13 @@ const FeedList = ({
             canceledFeeds.current.pop()
             userStorage.updateOTPLEventStatus(id, 'pending')
-            showErrorDialog('The payment could not be canceled at this time', 'CANCEL-PAYMNET-2')
+            showErrorDialog('The payment could not be canceled at this time. Please try again', ExceptionCode.E13)
```
add decorate

yaroslav-fedyshyn-nordwhale replied:
it already contains understandable error message

serdiukov-o-nordwhale commented on src/components/signup/SignupState.js:
```diff
@@ -602,7 +606,7 @@ const Signup = ({ navigation }: { navigation: any, screenProps: any }) => {
         return navigateWithFocus(nextRoute.key)
       } catch (e) {
         log.error('email verification failed unexpected:', e.message, e, { dialogShown: true })
```
add decorate

yaroslav-fedyshyn-nordwhale replied:
It already contains an understandable error message.

serdiukov-o-nordwhale commented on src/lib/API/api.js:
```diff
@@ -425,7 +425,7 @@ export class APIService {
       payload,
       () => log.info('addMauticContact success'),
-      ({ content }, xhr) => log.error('addMauticContact call failed:', content),
+      (response, xhr) => log.error('addMauticContact call failed:', '', null, { response }),
```
pls restore as it was before (remove your debug commit)

serdiukov-o-nordwhale commented:
> @serdiukov-o-nordwhale please check if I merged the analytic.js patchLogger changes correctly

yep, everything is fine

serdiukov-o-nordwhale commented:
@yaroslav-fedyshyn-nordwhale pls resolve conflicts

yaroslav-fedyshyn-nordwhale commented:
> backend error text that are displayed under signup form do not need the prefix
> something unexpected happened... see #2304

I reverted the server changes with error messages

serdiukov-o-nordwhale commented:
> > backend error text that are displayed under signup form do not need the prefix
> > something unexpected happened... see #2304
> 
> I reverted the server changes with error messages

as it was caused by  log.error() called on the server you could restore them back

sirpy commented on src/lib/logger/exceptions.js:
```diff
@@ -7,7 +7,9 @@ export const ExceptionCategory = {
-export const ExceptionCode = fromPairs(range(1, 13).map(code => [`E${code}`, code]))
+const MAX_EXCEPTION_CODE = 13
+
+export const ExceptionCode = fromPairs(range(1, MAX_EXCEPTION_CODE + 1).map(code => [`E${code}`, code]))
```
why not make a map of code->codeString
instead if using findKey which iterates over everything you just get ExceptionCode[code]
@serdiukov-o-nordwhale 
or why not simply codeString = `E${withCode}`

serdiukov-o-nordwhale replied:
this generates enum-like object

ExceptionCode = {
  E1: 1,
  E2: 2,
  // ...
}

which could be easy re-used

serdiukov-o-nordwhale replied:
i've removed findKey and used just `E${code}`

serdiukov-o-nordwhale commented on src/components/common/modal/ModalActionsByFeedType.js:
```diff
@@ -48,7 +49,7 @@ const ModalActionsByFeedType = ({ theme, styles, item, handleModalClose, navigat
       const { message } = exception
 
       userStorage.updateOTPLEventStatus(item.id, 'pending')
```
```suggestion
      decorate(exception, code)
      userStorage.updateOTPLEventStatus(item.id, 'pending')
```

serdiukov-o-nordwhale commented on src/components/common/modal/ModalActionsByFeedType.js:
```diff
@@ -48,7 +49,7 @@ const ModalActionsByFeedType = ({ theme, styles, item, handleModalClose, navigat
       userStorage.updateOTPLEventStatus(item.id, 'pending')
-      showErrorDialog('The payment could not be canceled at this time', code)
+      showErrorDialog('The payment could not be canceled at this time. Please try again.', code)
       log.error('cancel payment failed', message, exception, pickBy({ dialogShown: true, code, category }))
```
```suggestion
      log.error('cancel payment failed', message, exception, pickBy({ dialogShown: true, category }))
```

serdiukov-o-nordwhale replied:
let's send exception code like in the other places (e.g. as exception.code prop)

yaroslav-fedyshyn-nordwhale replied:
ok

serdiukov-o-nordwhale commented on src/components/dashboard/Claim.js:
```diff
@@ -117,20 +122,25 @@ const Claim = props => {
   const init = async () => {
-    //hack to make unit test pass, activityindicator in claim button cuasing
+    // hack to make unit test pass, activityindicator in claim button cuasing
     if (process.env.NODE_ENV !== 'test') {
```
is this for **unit** tests or for cypress ? if it's for cypress it should be updated to isE2ERunning imported from utils/platform

yaroslav-fedyshyn-nordwhale replied:
I checked it and found that it is for unit tests. Its solves 2 tests

serdiukov-o-nordwhale commented on src/components/signup/SignupState.js:
```diff
@@ -490,11 +491,15 @@ const Signup = ({ navigation }: { navigation: any, screenProps: any }) => {
-      log.error('New user failure', e.message, e, { dialogShown: true })
-      showSupportDialog(showErrorDialog, hideDialog, navigation.navigate, e.message)
+    } catch (exception) {
+      const { message } = exception
```
use decorate

serdiukov-o-nordwhale commented on src/components/dashboard/Dashboard.js:
```diff
@@ -570,7 +570,9 @@ const Dashboard = props => {
       } catch (exception) {
         const { message } = exception
-        const uiMessage = decorate(exception, ExceptionCode.E4)
+        const genericError = decorate(exception, ExceptionCode.E4)
+        const uiMessage =
+          message.indexOf('own payment') >= 0 ? "You can't withdraw your own payment link" : genericError
```
```suggestion
        let uiMessage = decorate(exception, ExceptionCode.E4)
        
        if (message.includes('own payment')) {
          uiMessage = message
        }
```
